src/app/routes/applications/new-role/components/role-field/role-field.component.spec.ts

```diff
@@ -0,0 +1,319 @@
+
+      expect(user[0].fieldType).toBe('text');
+      expect(user[0].label).toBe('Test');
+      expect(component.showFieldsForm).toBeFalsy();
+    });
+
+  });
+
+  describe('edit field button', () => {
+
+    it('edit button should open form to edit field with correct selected values', () => {
+      addFieldToRole();
+      editFieldAdded();
+
+      const fieldsValue = component.fieldsForm.getRawValue();
+
+      expect(component.showFieldsForm).toBeTruthy();
+      expect(fieldsValue.fieldType).toBe('text');
+      expect(fieldsValue.label).toBe('Test');
+    });
+
+    it('update button should replace old data and close form', () => {
+      addFieldToRole();
+      editFieldAdded();
+
+      selectFieldTypeAndLabel(1, 'Edit Label');
+
+      component.updateDataSourceEvent.subscribe(value => {
+        component.dataSource.data = [...value];
+        fixture.detectChanges();
+      });
+
+      const { updateField } = getSelectors(hostDebug);
+      updateField.nativeElement.click();
+      fixture.detectChanges();
+
+      const field = component.dataSource.data;
+      expect(field[0].fieldType).toBe('number');
+      expect(field[0].label).toBe('Edit Label');
+      expect(component.showFieldsForm).toBeFalsy();
+    });
+
+  });
+
+  describe('should be run', () => {
+
+    it('addField is invalid form', () => {
+      component.fieldsForm.setErrors({ incorrect: true });
+      spyOn(component, 'resetForm');
+      spyOn(component, 'updateDataSource');
+      component.addField();
+
+      expect(component.resetForm).not.toHaveBeenCalled();
+      expect(component.updateDataSource).not.toHaveBeenCalled();
+    });
+
+    it('back', () => {
+      spyOn(component.backEvent, 'emit');
+      component.back();
+
+      expect(component.backEvent.emit).toHaveBeenCalled();
+    });
+
+    it('proceedConfirmDetails', () => {
+      spyOn(component.proceedConfirmDetailsEvent, 'emit');
+      component.proceedConfirmDetails();
+
+      expect(component.proceedConfirmDetailsEvent.emit).toHaveBeenCalled();
+    });
+
+    it('resetForm', () => {
+      component.isEditFieldForm = true;
+      component.showFieldsForm = true;
+      spyOn(component.resetFormEvent, 'emit');
+
+      component.resetForm();
+
+      expect(component.isEditFieldForm).toBe(false);
+      expect(component.showFieldsForm).toBe(false);
+      expect(component.resetFormEvent.emit).toHaveBeenCalledWith(true);
+    });
+
+    it('updateDataSource', () => {
+      const data = {value: 'testData'};
+      spyOn(component.updateDataSourceEvent, 'emit');
+
+      component.updateDataSource(data);
+
+      expect(component.updateDataSourceEvent.emit).toHaveBeenCalledWith(data);
+    });
+
+    it('cancelAddField', () => {
+      spyOn(component, 'resetForm');
+
+      component.cancelAddField();
+
+      expect(component.resetForm).toHaveBeenCalled();
+    });
+
+    it('moveDown', () => {
+      component.dataSource.data = [1, 2, 3];
+      const i = 1;
```

Comment:
this variable is not needed i think. Or if you want it, then please name it better. For example `const indexToMove = 1`

Reply:
updated